os2phonebook_service/os2phonebook/datastore.py

```diff
@@ -0,0 +1,441 @@
+
+        Args:
+            search_field (str): Name of the document field to search in.
+            search_value (str): Arbitrary search string
+            size (int): Amount of documents to return
+            source_filter (list): List of field names to return
+
+        Returns:
+            dict: Elastic search query
+
+        """
+
+        query = {
+            "size": size,
+            "_source": {
+                "includes": list(source_filter)
+            },
+            "query": {
+                "match": {
+                    search_field: search_value
+                }
+            }
+        }
+
+        return query
+
+    def _query_match_phrase_prefix(self, search_field: str, search_value: str, size: int, source_filter: list) -> dict:
+        """Generate a query to match on the search prefix.
+
+        Unlike `match query` this is a broader search query,
+        matching on everything that `starts with`.
+
+        This is typically used as a second search attempt if there was no initial match.
+
+        For example:
+            search_type = "ho" will return only "holiday" but not "shoo"
+
+        Args:
+            search_field (str): Name of the document field to search in.
+            search_value (str): Arbitrary search string
+            size (int): Amount of documents to return
+
+            *args (args): List of field names to return
+
+        Returns:
+            dict: Elastic search query
+
+        """
+
+        query = {
+            "size": int(size),
+            "_source": {
+                "includes": list(source_filter)
+            },
+            "query": {
+                "match_phrase_prefix": {
+                    search_field: search_value
+                }
+            }
+        }
+
+        return query
+
+    def query_for_employee_by_name(self, name: str, fuzzy_search: bool) -> Tuple[str, dict]:
+        """Search query for an employee by his or her name.
+
+        Args:
+            name (str): Name of the document field to search in.
+            fuzzy_search (bool): Search wider if True
+
+        Returns:
+            Tuple[str, dict]: Index name (str) and Elastic search query (dict)
+
+        """
+
+        index = "employees"
+
+        source_filter = [
+            "uuid",
+            "name",
+            "addresses.PHONE",
+        ]
+
+        if fuzzy_search:
+            search_field = "givenname"
+            query = self._query_match_phrase_prefix(search_field, name, 15, source_filter)
+        else:
+            query = {
+                "_source": {
+                    "includes": list(source_filter)
+                },
+                "query": {
+                    "multi_match": {
+                        "query": str(name),
+                        "type": "cross_fields",
+                        "fields": ["surname", "givenname"],
+                        "tie_breaker": 0.3
+                    }
+                }
+            }
```

Comment:
I think we should have a short explanation of the reasoning behind the different queries - e.g. why we use a `multi_match` here.

Reply:
Better explanation for query types added.